### ISSUE TITLE
Loaders do not re-run after form submission if the form unmounts before the action finishes

### ISSUE THREAD

**Issue by camjackson** [remix-run/remix#3942]
### What version of Remix are you using?

1.6.7

### Steps to Reproduce

I have created a [SSCCE](http://www.sscce.org/) repo here: https://github.com/camjackson/remix-bug. It has the code to demonstrate the bug, steps to reproduce it, and a short video of the bad behaviour.

But in short:

1. Have a `<Form>` that auto-submits `onBlur`
2. Trigger a blur by clicking a `<Link>`, one which causes the `<Form>` to un-mount as part of the page navigation

### Expected Behavior

Once the action finishes, all the page's loaders should trigger, refreshing the data on the page.

### Actual Behavior

If the form un-mounts before the action finishes, the page's loaders do not trigger, so data on the page can get out-of-sync.

### Video

At 10 seconds, when navigating to person 2, the blur triggers and the person gets saved, but the loader for the list does not re-run, so the first name in the list does not change. At 12 seconds, you can see that the name in the list and the name on the form have become out of sync. At 13 seconds I refresh the page, which brings the list and the label back in sync. Though the input then goes back to the *original* value which is even more confusing but I think that's a different bug.

[bug.webm](https://user-images.githubusercontent.com/1930451/183238025-03fb565c-16e5-4683-b732-6baae30f5a56.webm)

**Comment by massimopalmieri** [comment]
I'm experiencing the same issue, have you found a solution?

**Comment by camjackson** [comment]
I think you already found it, but for future readers, there's more info and a suggested solution in [this discord thread](https://discord.com/channels/770287896669978684/1007600242490867712). I've yet to try the suggestion, will report back here when I have.

**Comment by camjackson** [comment]
Final update to this: the discord discussion went further and after some experimentation I ended up with a solution. See the thread for the full details, but in short, the solution was:

- `useFetcher` instead of `useSubmit`, because requests triggered by `useFetcher` are not cancelled upon navigation
- Move the `useFetcher` upwards to some component that will not be un-mounted when the navigation happens
- If it needs to go all the way up to a parent route, make it available to the child routes through [`<Route context />`](https://remix.run/docs/en/v1/api/remix#outlet-context-).

Something like:

```
// parentRoute.tsx
export default function ParentRoute() {
  const { submit } = useFetcher();
 
  return <Outlet context={{ submit }} />;
}

// childRoute.tsx
export default function ChildRoute() {
  const { submit } = useOutletContext();

  // These requests will never be cancelled as long as ParentRoute is mounted.
  const autoSave = e => submit(e.currentTarget, { replace: true });

  return (
    <Form onBlur={autoSave}>
      {/* Whatever */}
    </Form>
  );
}
```

Closing this as, while I don't like what remix is doing here, it doesn't seem like a bug per se. Just a gotcha.
